### PR TITLE
[EMS] Create dev tooling for config plugins

### PR DESCRIPTION
**Pull request by EvanBacon**
# Why

To create a [config plugin](https://github.com/expo/expo-cli/blob/master/packages/config-plugins/README.md) that automatically configures your native code, you have two options:

1. Create a `plugin` folder and write your plugin in TypeScript (Recommended).
2. Create an `app.plugin.js` and write the plugin in pure Node.js compliant JavaScript

Config plugins must be transpiled for Node.js (LTS) which is slightly different to native React modules which support features like the import/export keywords. This means we'll need two different `tsconfig.json` files and two different `src` (and `build`) folders.

This can quickly become complex, so we've created a system for easily targeting the plugin folder...

# How

- Created ems scripts for running `yarn build plugin`, `yarn clean plugin`, `yarn test plugin`, and `yarn lint plugin`
- Documented how to use scripts and setup a TypeScript plugin in an existing package

# Test Plan

- plugin tests are run using `yarn test plugin`.
- built a stripe plugin using the system to test the DX workflow. 
  - The plugin is applied in `native-component-list` when running `expo prebuild` then `yarn ios`
  - A `Payments` screen has been added to test the plugin and module (screen based on our [payments demo](https://github.com/expo/apple-pay-example/blob/master/App.js)).
- added plugins to `expo-tools check-packages` so they get run in CI. 
  - I'll request a review from @tsapeta on et after we decide on the EMS API.

### CODE REVIEW

File: apps/native-component-list/src/screens/PaymentsScreen.tsx [hunk 1]
```diff
@@ -0,0 +1,166 @@
...
+           */
+          requiredShippingAddressFields: ['all'],
+          /**
+           * An array of `ShippingMethod` objects that describe the supported shipping methods.
+           */
+          shippingMethods: [
+            {
+              id: 'zebra',
+              label: 'An actual Zebra 🦓',
+              detail: 'A real zebra will bring you the package 😳',
```
Comment (brentvatne) [inline apps/native-component-list/src/screens/PaymentsScreen.tsx:65]: great

File: packages/expo-module-scripts/README.md [hunk 2]
```diff
@@ -66,37 +67,110 @@ Running `yarn` will now run the `prepare` script, which generates any missing fi
 
 You should also add the following fields to your `package.json`:
 
+### 🔌 Config plugin
+
+To create a [config plugin](https://github.com/expo/expo-cli/blob/master/packages/config-plugins/README.md) that automatically configures your native code, you have two options:
+
+1. Create a `plugin` folder and write your plugin in TypeScript (Recommended).
+2. Create an `app.plugin.js` and write the plugin in pure Node.js compliant JavaScript
```
Comment (brentvatne) [inline packages/expo-module-scripts/README.md:75]:
```suggestion
2. Create an `app.plugin.js` in the project root and write the plugin in pure Node.js compliant JavaScript
```

File: packages/expo-module-scripts/README.md [hunk 3]
```diff
@@ -66,37 +67,110 @@ Running `yarn` will now run the `prepare` script, which generates any missing fi
...
+
+Config plugins must be transpiled for Node.js (LTS) which is slightly different to native React modules which support features like the import/export keywords. This means we'll need two different `tsconfig.json` files and two different `src` (and `build`) folders.
```
Comment (brentvatne) [inline packages/expo-module-scripts/README.md:77]:
```suggestion
Config plugins must be transpiled for compatibility with Node.js (LTS). The features supported in Node.js are slightly different to React Native modules, which support ES6 import/export keywords and JSX, for example. This means we'll need two different `tsconfig.json` files and two different `src` (and `build`) folders - one for the code that will execute in a React Native app and the other for the plugin that executes in Node.js.
```

File: packages/expo-module-scripts/tsconfig.plugin.json [hunk 4]
```diff
@@ -0,0 +1,7 @@
+{
+    "extends": "@tsconfig/node12/tsconfig.json",
```
Comment (brentvatne) [inline packages/expo-module-scripts/tsconfig.plugin.json:2]: technically we still support node 10 LTS also

File: packages/expo-payments-stripe/README.md [hunk 5]
```diff
@@ -21,6 +21,29 @@ For bare React Native projects, you must ensure that you have [installed and con
 expo install expo-payments-stripe
 ```
 
+- Include the config plugin in your `app.config.js` or `app.json`. Add the following props:
+  - `scheme` is the redirect URI that's used for returning to the app after an external payment. This should not match one of your existing schemes as it'll redirect to a special Activity on Android.
```
Comment (brentvatne) [inline packages/expo-payments-stripe/README.md:25]: might be better to give this a more specific name to make it clear in code that users shouldn't just pick a scheme that they're already using

File: packages/expo-payments-stripe/README.md [hunk 6]
```diff
@@ -21,6 +21,29 @@ For bare React Native projects, you must ensure that you have [installed and con
...
+  - `merchantId` an Apple ID that's required for making payments. This value should match the `merchantId` property used with `setOptionsAsync({ ... })` in your React code.
```
Comment (brentvatne) [inline packages/expo-payments-stripe/README.md:26]: `appleMerchantId`?

**Review by brentvatne — APPROVED** [review]

File: packages/expo-payments-stripe/plugin/src/withStripe.ts [hunk 7]
```diff
@@ -0,0 +1,194 @@
...
+  mainApplication,
+  scheme,
+}: {
+  mainApplication: AndroidConfig.Manifest.ManifestApplication;
+  scheme: string;
+}): AndroidConfig.Manifest.ManifestApplication {
+  if (Array.isArray(mainApplication.activity)) {
+    // Remove all Facebook CustomTabActivities first
+    mainApplication.activity = mainApplication.activity.filter(activity => {
+      return (activity.$ || {})['android:name'] !== CUSTOM_TAB_ACTIVITY;
```
Comment (ide) [inline packages/expo-payments-stripe/plugin/src/withStripe.ts:96]:
```suggestion
      return (activity.$ ?? {})['android:name'] !== CUSTOM_TAB_ACTIVITY;
```

File: packages/expo-payments-stripe/plugin/src/withStripe.ts [hunk 8]
```diff
@@ -0,0 +1,194 @@
...
+  }
+  // @ts-ignore: not on type yet
+  if (config.ios.scheme && !Array.isArray(config.ios.scheme)) {
+    // @ts-ignore: not on type yet
+    config.ios.scheme = [config.ios.scheme];
+  }
+  // @ts-ignore: not on type yet
+  config.ios.scheme.push(scheme);
+
+  // Append store kit
```
Comment (ide) [inline packages/expo-payments-stripe/plugin/src/withStripe.ts:127]:
```suggestion
  // Append StoreKit framework
```

File: packages/expo-payments-stripe/plugin/src/withStripe.ts (hunk 7 above)
Comment (ide) [inline packages/expo-payments-stripe/plugin/src/withStripe.ts:96]: What does the $ mean?

**Review by ide — CHANGES_REQUESTED** [review]
The tests for config plugins really make a difference

File: tools/src/Packages.ts [hunk 9]
```diff
@@ -72,6 +72,10 @@ export class Package {
     this.unimoduleJson = readUnimoduleJsonAtDirectory(rootPath);
   }
 
+  get hasPlugin(): boolean {
+    return fs.pathExistsSync(path.join(this.path, 'plugin'));
```
Comment (bbarthec) [inline tools/src/Packages.ts:76]: Isn't this check too optimistic? 🤔

File: tools/src/check-packages/checkPackageAsync.ts [hunk 10]
```diff
@@ -13,22 +13,29 @@ const { green } = chalk;
  */
 export default async function checkPackageAsync(
   pkg: Package,
-  options: ActionOptions
+  options: ActionOptions & { isPlugin?: boolean }
```
Comment (bbarthec) [inline tools/src/check-packages/checkPackageAsync.ts:16]:
```suggestion
  options: ActionOptions
```
Add this flag to the `ActionOptions` type in `./types` file 🥸

**Review by bbarthec — COMMENTED** [review]
Looks fine to me 👍 
Two small comments left 😉

File: packages/expo-payments-stripe/plugin/src/withStripe.ts (hunk 7 above)
Comment (EvanBacon) [reply packages/expo-payments-stripe/plugin/src/withStripe.ts:96]: In the XML package we use across the ecosystem `$` represents the head keys. 
Something like this: `<activity name="foo"></activity>` would become `{ $: { name: 'foo' } }`

File: packages/expo-payments-stripe/plugin/tsconfig.json [hunk 11]
```diff
@@ -0,0 +1,9 @@
+{
+    "extends": "../../expo-module-scripts/tsconfig.plugin",
```
[the remark this answers is not in the document]
Comment (EvanBacon) [reply packages/expo-payments-stripe/plugin/tsconfig.json:2]: this is based on the other tsconfig's in our monorepo

File: packages/expo-payments-stripe/README.md (hunk 6 above)
Comment (EvanBacon) [reply packages/expo-payments-stripe/README.md]: this matches the prop name used in the API, makes it a bit easier imo.

File: .gitignore [hunk 12]
```diff
@@ -133,3 +133,4 @@ docs/pages/versions/*/react-native/*.diff
 /packages/**/*.xcframework
 /packages/**/*.spec.json
 /packages/**/Info-generated.plist
+/packages/**/plugin/build/
```
Comment (ide) [inline .gitignore:136]: Sorry, I misspoke -- we want to commit (not .gitignore) the built plugin JS but want GitHub to collapse the built JS files during code review. This is handled in `.gitattributes`.

File: packages/expo-module-scripts/README.md [hunk 13]
```diff
@@ -66,37 +67,110 @@ Running `yarn` will now run the `prepare` script, which generates any missing fi
...
+```ts
+module.exports = require('expo-module-scripts/jest-preset-plugin');
+```
+
+Use the following scripts to interact with the plugin:
+
+- `yarn build plugin`: Build the plugin.
+- `yarn clean plugin`: Delete the `plugin/build` folder.
+- `yarn lint plugin`: Lint the `plugin/src` folder.
+- `yarn test plugin`: Alias for `npx jest --rootDir ./plugin --config ./plugin/jest.config.js`, uses the project jest preset if `plugin/jest.config.js` doesn't exist.
```
Comment (ide) [inline packages/expo-module-scripts/README.md:140]:
```suggestion
- `yarn test plugin`: Alias for `npx jest --rootDir ./plugin --config ./plugin/jest.config.js`, uses the project's Jest preset if `plugin/jest.config.js` doesn't exist.
```

File: tools/src/Packages.ts (hunk 9 above)
Comment (EvanBacon) [reply tools/src/Packages.ts:76]: good enough for our standard library

**Review by ide — APPROVED** [review]

**Pull request merged**